libraries/Microsoft.Bot.Builder.Dialogs/DialogExtensions.cs

```diff
@@ -90,6 +90,16 @@ public static async Task RunAsync(this Dialog dialog, ITurnContext turnContext,
             }
         }
 
+        // We should only cancel the current dialog stack if the EoC activity is coming from a parent (a root bot or another skill).
+        // When the EoC is coming back from a child, we should just process that EoC normally through the 
+        // dialog stack and let the child dialogs handle that.
+        private static bool IsEocComingFromParent(ITurnContext turnContext)
+        {
+            // To determine the direction we check callerId property which is set to the parent bot
+            // by the BotFrameworkHttpClient on outgoing requests.
+            return !string.IsNullOrWhiteSpace(turnContext.Activity.CallerId);
```

Comment:
Do we need to check the `Activity.CallerId` value or is it just removed by the skill by default?

For others' context, the `CallerId` is set to the `fromBotId` by the BotFrameworkHttpClient class [here](https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/BotFrameworkHttpClient.cs#L103).

Reply:
Nevermind, it's not intended to be sent over the wire per the spec. 

[swagger](https://github.com/microsoft/botbuilder-dotnet/blob/8efdbd723f0ceaa9ece353ca359ae4a44576794b/libraries/Swagger/ConnectorAPI.json#L952-L955):

```json
"callerId": {
  "description": "A string containing an IRI identifying the caller of a bot. This field is not intended to be transmitted\r\nover the wire, but is instead populated by bots and clients based on cryptographically verifiable data\r\nthat asserts the identity of the callers (e.g. tokens).",
  "type": "string"
},
```

Reply:
The Activity.CallerId is only set by BotFrameworkHttpClient on the way out from a parent to a skill, it is never set in activities sent back to bot the called the skill. 
This check allows us to know in what direction is the activity going (we should only cancelDialogs if the EoC comes from a parent and process it normally if it comes from a child)
At this point in the processing, I couldn't find other way of finding your if the EoC activity is here because it was received from the parent bot or if it is here because it was rerouted to the pipeline by SkillHandler in [ProcessActivityAsync](https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Builder/Skills/SkillHandler.cs#L192)

Reply:
`$"urn:botframework:aadappid:{fromBotId}"`